### PR TITLE
feat(EMS-1172): No PDF - CRON jobs

### PR DESCRIPTION
**Pull request by Zainzzkk**
## Introduction :pencil2:
This PR adds the CRON jobs for application and accounts:
* `updateUnverifiedAccountsJob` - updates accounts which are unverified for over 24hrs to isInactive
* `updateInactiveApplicationsJob` - updates applications which have not been updated for over 30days to Abandoned

## Resolution :heavy_check_mark:
* Added CRON jobs for application and accounts - `updateUnverifiedAccountsJob` and `updateInactiveApplicationsJob`
* Added types for CRON jobs
* Added middleware for CRON jobs to run
* Added `extendHttpServer` to `keystone.ts` to run CRON jobs when API starts
* Added env to env.sample and docker file
* Added unit tests
* Added /cron directory to api docker file

## Miscellaneous :heavy_plus_sign:
* Updated documentation for get-unverified-accounts
* Added ignore to `.prettierignore` for `helpers/cron/scheduler.ts`
* Updated `.cspell.json`

### CODE REVIEW

File: .prettierignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 src/ui/public
 src/ui/generated_reports
+src/api/helpers/cron/scheduler.ts
```
Comment (ttbarnes) [inline .prettierignore:3]: Do we need this - I can have a look at the linting issue, try to fix it

File: docker-compose.yml [hunk 2]
```diff
@@ -62,6 +62,8 @@ services:
       ORDNANCE_SURVEY_API_KEY:
       ORDNANCE_SURVEY_API_URL:
       LOSS_PAYEE_ENCRYPTION_KEY:
+      UNVERIFIED_ACCOUNT_SCHEDULE:
+      INACTIVE_APPLICATION_SCHEDULE:
```
Comment (ttbarnes) [inline docker-compose.yml:66]: Should we prefix these with `CRON_`? Or maybe 

```js
CRON_SCHEDULE_UNVERIFIED_ACCOUNT
CRON_SCHEDULE_INACTIVE_APPLICATION
```

File: src/api/cron/account/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { accountCronSchedulerJobs } from '.';
+import { updateUnverifiedAccountsJob } from './unverified-account-cron-job';
+
+describe('cron/account', () => {
+  describe('accountCronSchedulerJobs', () => {
+    it('should have an array of account cron tasks', () => {
```
Comment (ttbarnes) [inline src/api/cron/account/index.test.ts:6]: cron jobs* instead of tasks?

File: src/api/cron/account/unverified-account-cron-job.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import dotenv from 'dotenv';
+import updateUnverifiedAccounts from '../../helpers/update-unverified-accounts';
+import { CronSchedulerJob } from '../../types';
+
+dotenv.config();
+
+const { UNVERIFIED_ACCOUNT_SCHEDULE } = process.env;
+
+const unverifiedAccountSchedule = UNVERIFIED_ACCOUNT_SCHEDULE as string;
```
Comment (ttbarnes) [inline src/api/cron/account/unverified-account-cron-job.ts:9]: Maybe can do `String(UNVERIFIED_ACCOUNT_SCHEDULE)`

File: src/api/cron/account/unverified-account-cron-job.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * CronSchedulerJob object which contains:
+ * cronExpression: schedule on which the cron job runs - '* * * * * *'
+ * description: description of cronjob
+ * task: function which cron job runs - updateUnverifiedAccounts
+ */
+export const updateUnverifiedAccountsJob: CronSchedulerJob = {
+  cronExpression: unverifiedAccountSchedule,
+  description: 'Update unverified accounts (over 24hrs) to isInactive',
+  task: updateUnverifiedAccounts,
+};
```
Comment (ttbarnes) [inline src/api/cron/account/unverified-account-cron-job.ts:22]: export default?

Can the description be in constants?

File: src/api/cron/application/inactive-application-cron-job.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * CronSchedulerJob object which contains:
+ * cronExpression: schedule on which the cron job runs - '* * * * * *'
+ * description: description of cronjob
+ * task: function which cron job runs - updateInactiveApplications
+ */
+export const updateInactiveApplicationsJob: CronSchedulerJob = {
+  cronExpression: inactiveApplicationSchedule,
+  description: 'Update inactive applications (over 30 days) to Abandoned',
+  task: updateInactiveApplications,
+};
```
Comment (ttbarnes) [inline src/api/cron/application/inactive-application-cron-job.ts:22]: Same as above

File: src/api/cron/application/index.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import { CronSchedulerJob } from '../../types';
+import { updateInactiveApplicationsJob } from './inactive-application-cron-job';
+
+/**
+ * applicationCronSchedulerJobs
+ * Array of CronSchedulerJobs: cron jobs to schedule and run
+ */
+export const applicationCronSchedulerJobs: CronSchedulerJob[] = [updateInactiveApplicationsJob];
```
Comment (ttbarnes) [inline src/api/cron/application/index.ts:8]: Same as above

File: src/api/cron/index.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * cronJobScheduler
+ * Validates and sets up cron jobs to run
+ * @param {CronSchedulerJob[]} jobs: cron jobs to schedule to run
+ * @param {Context} KeystoneJS context API
+ */
+export const cronJobScheduler = (jobs: CronSchedulerJob[], context: Context) => {
+  jobs.forEach((job) => {
+    const { cronExpression, description, task } = job;
```
Comment (ttbarnes) [inline src/api/cron/index.ts:14]: Could we have a `Job` type? Instead of `CronSchedulerJob[]`, the `job` in the `forEach` could also consume this

Comment (ttbarnes) [reply src/api/cron/index.ts:14]: `export default`?

File: src/api/keystone.ts [hunk 9]
```diff
@@ -54,6 +55,13 @@ export default withAuth(
           app.use(rateLimiter);
         }
       },
+      extendHttpServer: (httpServer, commonContext) => {
+        /**
+         * runs cron jobs based on their schedule
+         * starts running when API is started
+         */
+        cronJobs(commonContext);
```
Comment (ttbarnes) [inline src/api/keystone.ts:63]: Could we rename `commonContext` to just `context`?

File: src/api/middleware/cron/index.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+ * cronJobs
+ * Middleware to run cron jobs
+ * @param {Context} KeystoneJS context API
+ */
+const cronJobs = (context: Context) => {
+  console.info('Running cron jobs');
+  // account cron jobs
+  cronJobScheduler(accountCronSchedulerJobs, context);
+  // application cron jobs
+  cronJobScheduler(applicationCronSchedulerJobs, context);
```
Comment (ttbarnes) [inline src/api/middleware/cron/index.ts:16]: Could we add a space before each comment so it's easier to read?

Also I think this should be moved to `helpers` or a new `cron` directory, as this is not middleware anymore

File: src/api/types/cron/index.ts [hunk 11]
```diff
@@ -1,3 +1,35 @@
...
+  /**
+   * A description of the job
+   */
+  description: string;
+  /**
+   * The task to be run by the job
+   */
+  task: CronSchedulerJobTask;
+};
+
```
Comment (ttbarnes) [inline src/api/types/cron/index.ts:32]: Lots of eslint disabling in here?

Comment (Zainzzkk) [reply src/api/types/cron/index.ts:32]: they are required.  Same in DTFS.  ESLint does not allow Function declaration but it is required.  Also the variables need to be declared but can't be used in the types file so need a disable too

File: .prettierignore (hunk 1 above)
Comment (Zainzzkk) [reply .prettierignore:3]: Sure, if it can be fixed I can remove this

Comment (ttbarnes) [reply .prettierignore:3]: Our eslint config is slightly different to DTFS.

I got this the `cron/scheduler.ts` file passing with prettier by updating our `eslintrc.js` - in the `extends` array;

Replacing these:

```js
'plugin:prettier/recommended',
'prettier/prettier',
```

With this (like DTFS):

```js
pretttier
```

...And also removing the `prettier/prettier` config from `rules`.

I'm not sure about any other consequences of this though, need to investigate.

**Review by ttbarnes — COMMENTED** [review]
Can we add to the PR description some job descriptions, instead of just "cron jobs for application and accounts"

File: src/api/cron/account/unverified-account-cron-job.test.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import dotenv from 'dotenv';
+import updateUnverifiedAccountsJob from './unverified-account-cron-job';
+import updateUnverifiedAccounts from '../../helpers/update-unverified-accounts';
+import { UPDATE_UNVERIFIED_ACCOUNT_DESCRIPTION } from '../../constants';
```
Comment (ttbarnes) [inline src/api/cron/account/unverified-account-cron-job.test.ts:4]: Maybe `CRON_DESCRIPTION` and entity prefixes would be nicer  e.g

```js
CRON_DESCRIPTION_ACCOUNT_UPDATE_UNVERIFIED
CRON_DESCRIPTION_APPLICATION_UPDATE_INACTIVE
``

**Review by ttbarnes — APPROVED** [review]

**Pull request merged**